src/compute/src/sink/copy_to_s3_oneshot.rs

```diff
@@ -67,68 +81,266 @@ where
+                    }
+                }
+                AsyncEvent::Progress(frontier) => {
+                    if PartialOrder::less_equal(&up_to, &frontier) {
+                        match uploader.finish_ongoing_upload().await {
+                            Ok(()) => {}
+                            Err(e) => {
+                                send_response(CopyToResponse::Error(e.to_string()));
+                                return;
+                            }
+                        }
+                        // We are done, send the final count.
+                        send_response(CopyToResponse::RowCount(row_count));
+                        return;
+                    }
                 }
             }
         }
     });
 }
 
+// Required state to upload batches to S3
+struct CopyToS3Uploader {
+    // The output description.
+    desc: RelationDesc,
+    // Params to format the data.
+    format: CopyFormatParams<'static>,
+    // The index of the next file.
+    file_index: usize,
+    // The prefix for the file names.
+    file_name_prefix: String,
+    // The bucket prefix batches should be uploaded to.
+    bucket: String,
+    //The path prefix.
+    path_prefix: String,
+    // The desired batch size. When the buffer exceeds this value a put operation will be
+    // performed.
+    max_file_size: u64,
+    // The aws sdk config.
+    sdk_config: SdkConfig,
+    // Keeping the uploader in an `Option` to later take owned value.
+    single_file_uploader: Option<S3MultiPartUploader>,
+    // temporary buffer
+    buf: Vec<u8>,
+}
+
+impl CopyToS3Uploader {
+    fn new(
+        sdk_config: SdkConfig,
+        sink_connection: CopyToS3OneshotSinkConnection,
+        file_name_prefix: String,
+    ) -> CopyToS3Uploader {
+        let (bucket, path_prefix) = Self::extract_s3_bucket_path(&sink_connection.prefix);
+        CopyToS3Uploader {
+            desc: sink_connection.desc,
+            sdk_config,
+            format: sink_connection.format,
+            file_name_prefix,
+            bucket,
+            path_prefix,
+            max_file_size: sink_connection.max_file_size,
+            file_index: 0,
+            single_file_uploader: None,
+            buf: Vec::new(),
+        }
+    }
+
+    async fn start_new_file_upload(&mut self) -> Result<(), anyhow::Error> {
+        self.file_index += 1;
+        let file_path = self.file_path();
+        let bucket = self.bucket.clone();
+        info!(
+            "starting upload at bucket: {}, file {}",
+            &bucket, &file_path
+        );
+        let uploader = S3MultiPartUploader::try_new(
+            &self.sdk_config,
+            bucket,
+            file_path,
+            S3MultiPartUploaderConfig {
+                part_size_limit: ByteSize::mib(10).as_u64(),
+                file_size_limit: self.max_file_size,
+            },
+        )
+        .await?;
+        self.single_file_uploader = Some(uploader);
+        Ok(())
```

Comment:
```suggestion
        self.single_file_uploader.insert(uploader)
```

Reply:
not using `insert` since we don't need to return the mutable borrow.